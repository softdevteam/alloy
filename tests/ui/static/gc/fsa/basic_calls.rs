#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}

impl<T: Debug> Drop for Wrapper<T> {
    fn drop(&mut self) {
        foo(self);
        bar(self);
        baz(self);
    }
}

#[inline(never)]
fn foo<T: Debug>(x: &Wrapper<T>) {
    use_val(&x.0); // should fail
}

#[inline(never)]
fn bar<T: Debug>(x: &Wrapper<T>) {
    foo(x);
}

#[inline(never)]
fn baz<T: Debug>(x: &Wrapper<T>) {
    use_val(&x.0); // should fail
    use_val(&x); // should pass (no projection)
}

fn main() {
    Gc::new(Wrapper(FinalizerUnsafeU8Wrapper(1)));
    //~^ ERROR: The drop method for `Wrapper<FinalizerUnsafeU8Wrapper>` cannot be safely finalized.
    //~| ERROR: The drop method for `Wrapper<FinalizerUnsafeU8Wrapper>` cannot be safely finalized.

    Gc::new(Wrapper(Wrapper(1)));
}
