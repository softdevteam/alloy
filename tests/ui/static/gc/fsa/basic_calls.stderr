error: The drop method for `Wrapper<FinalizerUnsafeU8Wrapper>` cannot be safely finalized.
  --> $DIR/basic_calls.rs:32:13
   |
LL |     use_val(&x.0); // should fail
   |             ----
   |             |
   |             a finalizer cannot safely use this `FinalizerUnsafeU8Wrapper`
   |             from a drop method because it does not implement `FinalizerSafe`.
...
LL |     Gc::new(Wrapper(FinalizerUnsafeU8Wrapper(1)));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<Wrapper<FinalizerUnsafeU8Wrapper>>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.

error: The drop method for `Wrapper<FinalizerUnsafeU8Wrapper>` cannot be safely finalized.
  --> $DIR/basic_calls.rs:32:13
   |
LL |     use_val(&x.0); // should fail
   |             ----
   |             |
   |             a finalizer cannot safely use this `FinalizerUnsafeU8Wrapper`
   |             from a drop method because it does not implement `FinalizerSafe`.
...
LL |     Gc::new(Wrapper(FinalizerUnsafeU8Wrapper(1)));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<Wrapper<FinalizerUnsafeU8Wrapper>>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.

error: aborting due to 2 previous errors

